### ISSUE TITLE
習題 12 : 請用梯度下降法之類的優化算法，驗證 cross_entropy(p, q) 的最低點是 q=p

### ISSUE THREAD

**Issue by ccckmit** [ccc114a/_alg#14]
請用優化算法，驗證 cross_entropy(p, q) 的最低點是 q=p

```py
def cross_entropy(p,q):
    r = 0
    for i in range(len(p)):
        r += p[i]*log2(1/q[i])
    return r
```

範例

1. p=[1/2, 1/4, 1/4], q = [1/3,1/3,1/3] , 可調 q, 不能調 p, 驗證交叉熵的最小值是 entropy(p), 也就是 cross_entropy(p,p)

障礙

1. 使用梯度下降法調整 q 時，可能會造成 q 的加總不是 1 ，所以每次調整之後，應該還要再 normalize, 使 q 的總和為 1 (歸一化）（實驗結果這個方法失敗）
2. 使用 爬山演算法，成功！

**Comment by ccckmit** [comment]
參考答案，我使用 Google AI Studio ，用梯度下降法無法找到最低點（因為最後還要歸一化），所以 AI 改用爬山演算法，就可以了

* [Google AI Studio 優化 cross_entropy(p,q) 的對話](https://aistudio.google.com/app/prompts?state=%7B%22ids%22:%5B%221mkABwEQ4B6vkG04x7mGoviC6NRj7XtyC%22%5D,%22action%22:%22open%22,%22userId%22:%22111605452542833299008%22,%22resourceKeys%22:%7B%7D%7D&usp=sharing)


```py
import math
import numpy as np

np.set_printoptions(precision=4, suppress=True)

# ---------------------------------------------------------
# 1. 定義 Cross Entropy
# ---------------------------------------------------------
def log2(x):
    # 防止 log(0) 錯誤
    return math.log(max(x, 1e-15), 2)

def cross_entropy(p, q):
    r = 0
    for i in range(len(p)):
        r += p[i] * log2(1 / q[i])
    return r

def entropy(p):
    return cross_entropy(p, p)

# ---------------------------------------------------------
# 2. 隨機爬山演算法 (Random Hill Climbing)
# ---------------------------------------------------------
def random_hill_climbing(loss_func, p_target, q_init, max_loops=10000):
    
    # 複製初始狀態
    q = q_init.copy()
    current_loss = loss_func(p_target, q)
    
    # 初始搜尋半徑 (步長)
    step_size = 0.5
    
    print(f"Start Random Search...")
    print(f"Initial q: {q}, Loss: {current_loss:.5f}")
    
    for i in range(max_loops):
        # 1. 產生隨機擾動 (Random Noise)
        # 產生一個跟 q 一樣長的隨機向量，數值在 -1 到 1 之間
        noise = np.random.uniform(-1, 1, size=q.shape)
        
        # 2. 試探新的點
        q_try = q + (noise * step_size)
        
        # --- 解決障礙：約束處理 ---
        # (A) 確保所有數值 > 0 (取絕對值或 max)
        q_try = np.abs(q_try) + 1e-10 
        
        # (B) 歸一化：確保總和為 1
        q_try = q_try / np.sum(q_try)
        
        # 3. 計算新 Loss
        new_loss = loss_func(p_target, q_try)
        
        # 4. 決策：如果新 Loss 比較小，就移動過去 (貪婪策略)
        if new_loss < current_loss:
            q = q_try
            current_loss = new_loss
        
        # 5. 動態調整步長 (Learning Rate Decay)
        # 隨著時間過去，縮小搜尋範圍，讓結果更精確
        if i % 1000 == 0:
            step_size *= 0.8 # 每次縮小 20%
            print(f"{i:05d}: Loss={current_loss:.5f} q={q} step={step_size:.5f}")
            
    return q

# ---------------------------------------------------------
# 3. 主程式驗證
# ---------------------------------------------------------
if __name__ == "__main__":
    # 目標分佈 p
    p = np.array([1/2, 1/4, 1/4])
    
    # 初始猜測 q
    q_start = np.array([1/3, 1/3, 1/3])
    
    print(f"Target p: {p}")
    print(f"Target Min Loss (Entropy): {entropy(p):.5f}\n")
    
    # 執行優化
    q_final = random_hill_climbing(cross_entropy, p, q_start, max_loops=20000)
    
    print("-" * 60)
    print("Final Result:")
    print(f"Optimized q : {q_final}")
    print(f"Target    p : {p}")
    print(f"Final Loss  : {cross_entropy(p, q_final):.5f}")
    
    # 驗證
    error = np.sum(np.abs(q_final - p))
    if error < 0.01:
        print("\n✅ 驗證成功：在不使用梯度的情況下，隨機搜尋法找到了 q=p")
    else:
        print("\n⚠️ 未完全收斂")
```

執行結果

```sh
(base) teacher@teacherdeiMac 01-entropy % python hc_cross_entropy.py
Target p: [0.5  0.25 0.25]
Target Min Loss (Entropy): 1.50000

Start Random Search...
Initial q: [0.3333 0.3333 0.3333], Loss: 1.58496
00000: Loss=1.58496 q=[0.3333 0.3333 0.3333] step=0.40000
01000: Loss=1.50050 q=[0.507  0.2543 0.2387] step=0.32000
02000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.25600
03000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.20480
04000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.16384
05000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.13107
06000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.10486
07000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.08389
08000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.06711
09000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.05369
10000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.04295
11000: Loss=1.50000 q=[0.4988 0.2504 0.2507] step=0.03436
12000: Loss=1.50000 q=[0.5001 0.25   0.2499] step=0.02749
13000: Loss=1.50000 q=[0.5001 0.25   0.2499] step=0.02199
14000: Loss=1.50000 q=[0.5    0.25   0.2499] step=0.01759
15000: Loss=1.50000 q=[0.5    0.25   0.2499] step=0.01407
16000: Loss=1.50000 q=[0.5    0.25   0.2499] step=0.01126
17000: Loss=1.50000 q=[0.5    0.25   0.2499] step=0.00901
18000: Loss=1.50000 q=[0.5    0.25   0.2499] step=0.00721
19000: Loss=1.50000 q=[0.4999 0.25   0.2501] step=0.00576
------------------------------------------------------------
Final Result:
Optimized q : [0.5    0.2499 0.25  ]
Target    p : [0.5  0.25 0.25]
Final Loss  : 1.50000

✅ 驗證成功：在不使用梯度的情況下，隨機搜尋法找到了 q=p
```

**Comment by tonytomtody** [comment]
## 111210504 蔡瑞哲
[HW12](https://github.com/tonytomtody/_alg/blob/main/week13/cross-en.py)

**Comment by CYR1225** [comment]
資工三 陳奕儒 111210546
[HW12](https://github.com/tonytomtody/_alg/blob/main/week13/cross-en.py)

**Comment by 111210511** [comment]
資工三 李昱欣 111210511 
[hw12](https://github.com/111210511/_alg/blob/main/hw12)

**Comment by Jonathan-es** [comment]
資工三 111210554 陳家盛
[hw12](https://github.com/Jonathan-es/_alg/blob/main/hw12.py)

**Comment by edwimtiongyuanxie** [comment]
111210550 資工三 張源熙
[hw12](https://github.com/edwimtiongyuanxie/_alg/blob/main/hw12)

**Comment by 1100jimjim** [comment]
111210523 資工三 蘇湧竣
[HW12](https://github.com/1100jimjim/_alg/tree/main/HW12)

**Comment by shanghua0203** [comment]
111210505 資工三 吳尚樺
-
[HW12](https://github.com/shanghua0203/_alg/tree/main/HW12)

**Comment by peterwang0329** [comment]
#### 111210513 資工三 汪章貴

[HW12](https://github.com/peterwang0329/_alg/tree/main/hw/hw12)

**Comment by linpeic** [comment]
111210522 資工三 林沛欣
[HW12](https://github.com/linpeic/_alg/blob/main/HW12/q_fire.py)

**Comment by Erkmrcl17** [comment]
**111210559 資工三 李麗恩**
[HW12](https://github.com/Erkmrcl17/_alg/blob/main/HW12.py)

**Comment by codewhight** [comment]
111210510 資工三 林彥廷
[h12](https://github.com/codewhight/_alg/blob/main/h12/qp.py)

**Comment by jerry92916** [comment]
資工三 111210529 陳宏傑
[HW12](https://github.com/jerry92916/_alg/tree/main/HW12)

**Comment by hmzhen12** [comment]
資工三 111210548 胡眉真
[Homework 12](https://github.com/hmzhen12/_alg/tree/main/HW12)

**Comment by julianalidya** [comment]
111210552 資工三 林小蓮
[HW12](https://github.com/julianalidya/_alg/blob/main/HW%2012/cross-en.py)

**Comment by daibao87** [comment]
資工三 111210507 林育民
[hw12](https://github.com/daibao87/_alggg/tree/main/hw11)

**Comment by brain363663-cloud** [comment]
電機三 111210228 謝承恩
[HW12](https://github.com/brain363663-cloud/_alg/blob/b19307378f16d913079e1832ea7b843ab23d37ce/HW12/12.py)

**Comment by Huang9315** [comment]
資工四111110520黃志仁
[hw12](https://github.com/Huang9315/_alg/blob/main/%E6%9C%9F%E6%9C%AB%E6%95%B4%E7%90%86/hw12/Cross%20Entropy.py)

**Comment by cindycallista** [comment]
111210555 資工三 楊莉思
[hw12](https://github.com/cindycallista/_alg/tree/main/hw12)

**Comment by CH623** [comment]
資工三 111210515 趙興華
[hw12](https://github.com/CH623/_alg/blob/main/HW/HW12.py)

**Comment by Sakuraebi128** [comment]
111210541 資工三 黃瓘閎
[Hw12](https://github.com/Sakuraebi128/_alg/tree/main/Hw12)

**Comment by GuanCheng0430** [comment]
111210527 資工三 王冠程
[HW12](https://github.com/GuanCheng0430/ccc114a/blob/main/h12)
